test/vfd_swmr_group_writer.c

```diff
@@ -310,27 +315,87 @@ main(int argc, char **argv)
     if (s.file == badhid)
         errx(EXIT_FAILURE, writer ? "H5Fcreate" : "H5Fopen");
 
-    block_signals(&oldsigs);
+    /* Writer creates two named pipes(FIFO) to coordinate two-way communication */
+    if (writer) {
+        if (HDmkfifo(fifo_writer_to_reader, 0600) < 0)
+            errx(EXIT_FAILURE, "HDmkfifo");
+
+        if (HDmkfifo(fifo_reader_to_writer, 0600) < 0)
+            errx(EXIT_FAILURE, "HDmkfifo");
+    }
+
+    /* Both the writer and reader open the pipes */
+    if ((fd_writer_to_reader = HDopen(fifo_writer_to_reader, O_RDWR)) < 0)
+        errx(EXIT_FAILURE, "fifo_writer_to_reader open failed");
+
+    if ((fd_reader_to_writer = HDopen(fifo_reader_to_writer, O_RDWR)) < 0)
+        errx(EXIT_FAILURE, "fifo_reader_to_writer open failed");
 
     if (writer) {
+        /* Writer tells reader to start */
+        if (HDwrite(fd_writer_to_reader, &notify, sizeof(int)) < 0)
+            err(EXIT_FAILURE, "write failed");
+
         for (step = 0; step < s.nsteps; step++) {
-            dbgf(2, "step %d\n", step);
+            dbgf(2, "writer: step %d\n", step);
+
             write_group(&s, step);
-            nanosleep(&s.update_interval, NULL);
+
+            /* At communication interval, notifies the reader and waits for its response */
+            if (step % s.csteps == 0) {
+                /* Bump up the value and notify the reader */
+                notify++;
+                if (HDwrite(fd_writer_to_reader, &notify, sizeof(int)) < 0)
+                    err(EXIT_FAILURE, "write failed");
+
+                /* During the wait, writer makes repeated HDF5 API calls
+                 * to trigger EOT at approximately the correct time */
+                for(i = 0; i < config.max_lag + 1; i++) {
+                    decisleep(config.tick_len);
+                    H5Aexists(s.file, "nonexistent");
+                }
+
+                /* Receive the same value from the reader and verify it */
+                verify++;
+                if (HDread(fd_reader_to_writer, &notify, sizeof(int)) < 0)
+                    err(EXIT_FAILURE, "read failed");
+
+                if (notify != verify)
+                    errx(EXIT_FAILURE, "received message %d, expecting %d", notify, verify);
+            }
         }
     } else {
-        for (step = 0; step < s.nsteps;) {
-            dbgf(2, "step %d\n", step);
-            if (verify_group(&s, step))
-                step++;
-            nanosleep(&s.update_interval, NULL);
-        }
-    }
+        /* Start to verify group creation after receiving the writer's notice */
```

Comment:
Expanding on the above, what does the writer's notice mean?  That group creation has started, finished, or ??

Reply:
Done.